src/Compilers/Core/Portable/CommandLine/CommonCompiler.CompilerEmitStreamProvider.cs

```diff
@@ -40,15 +40,71 @@ public override Stream CreateStream(DiagnosticBag diagnostics)
 
                 try
                 {
-                    return _streamToDispose = _compiler.FileOpen(_filePath, PortableShim.FileMode.Create, PortableShim.FileAccess.ReadWrite, PortableShim.FileShare.None);
+                    try
+                    {
+                        return OpenFileStream();
+                    }
+                    catch (IOException e)
+                    {
+                        // Other process is reading the file preventing us to write to it.
+                        // We attempt to rename and delete the file in case the reader opened it with FileShare.Delete flag that
+                        // allows the file to be deleted by other processes.
+                        //
+                        // Note that if the file is marked "readonly" or the current user doesn't have sufficient privileges
+                        // the exception thrown is UnauthorizedAccessException, not IOException, so we won't attempt to delete the file.
+
+                        try
+                        {
+                            const int eWin32SharingViolation = unchecked((int)0x80070020);
+
+                            if (PathUtilities.IsUnixLikePlatform)
+                            {
+                                // Unix & Mac are simple: just delete the file in the directory. 
+                                // The memory mapped content remains available for the reader.
+                                PortableShim.File.Delete(_filePath);
+                            }
+                            else if (e.HResult == eWin32SharingViolation)
+                            {
+                                // On Windows File.Delete only marks the file for deletion, but doens't remove it from the directory.
+                                var newFilePath = Path.Combine(Path.GetDirectoryName(_filePath), Guid.NewGuid().ToString() + "_" + Path.GetFileName(_filePath));
```

Comment:
Why move the file within the directory vs. say moving it to `%TEMP%`?  

This could potentially impact items like deterministic build systems.  It would count as a new potential output of the compiler.  Given it has a generated name it wouldn't be predictable.